### PR TITLE
New testing proposal

### PR DESCRIPTION
**Pull request by vzarytovskii**
As part of our initiative of F# tests reworking, I would like to submit a proposal PR for new tests structure, frameworks, goals, and some initial prerequisites implementation.

I would really appreciate any feedback and additions.

# F# Testing proposal

## Why do we test

* To prevent regressions (behavioral, performance).
* To have a quicker debug feedback (thus, find problems quicker).
* To verify conformance to language spec (API contract testing).
* To have IL verification (both read and write).
* To have a quicker design feedback.
* To document behavior.

## Goals

* Use one standardized testing framework across all of test projects, and get rid of custom old solutions (FSharpQA and Cambridge suites).
* Have tests restructured the way, that they are easy to discover.
* Have tests building and running on all supported platforms (Windows, macOS and Linux) and different frameworks (with exceptions when this is not applicable).
* Make it easy to run tests using standard .NET instruments (dotnet cli, test explorer, etc.).
* Leverage standard .NET testing platform and use all its benefits, suck as live unit testing, code coverage collecting, dead code elimination, etc.

## Framework for testing

The following test frameworks and libraries will be used for new test projects **[xUnit Test Framework](https://xunit.net/), [FluentAssertions](https://fluentassertions.com/) (+ [FsUnit](https://fsprojects.github.io/FsUnit/) and [FsCheck](https://github.com/fscheck/FsCheck) when needed)**. All existing NUnit test suites will be migrated to xUnit.

**Justification:**

* **xUnit** is an extensible, TDD adherent, testing framework, which was successfully adopted by many .NET engineering teams, including Roslyn, AspNetCore, EFcore, etc, has a "cleaner" approach for writing test suites (i.e. class constructor for setup, implementing IDisposable for teardown, as oppose to custom attributes). More info [here](https://xunit.net/docs/comparisons).
* **FluentAssertions** makes it easier to write scoped assertions, provides better error messages.

**Alternatives:** NUnit, MSBuild, Expecto

#### New tests should be grouped based on two factors: test type (1) + test category and subcategory (2)

1. **Test type**:
**Determines what type of test is it:**
   * __Functional tests__:
        * __Unit Tests__: a lightweight testing for smaller modules, functions, etc.
          * __Examples__: Testing individual parts/functions of lexer, parser, syntax tree, standard library modules, etc.
          * __Subgroups__: there should be a separation between testing private and public parts of each module (i.e. compiler tests for private and public API should be in separate test projects).
        * __Component Tests__: testing for bigger parts of compiler.
          * __Examples__: Tests for the compiler components as whole, such as Code generation, IL Generation, Compiler optimizations, Type Checker, Type Providers, Conformance, etc.
        * __Integration and End2End Tests__: testing of F# compiler & tooling integration, as well as e2e experiences.
          * __Examples__: VS Integration, .NET Interactive integration, LSP integration. Integration with dotnet CLI, project creation, building, running.
   * __Non-functional tests__:
        * __Load and Stress Tests__: testing for high level modules/components to understand peak performance and potentially catch any performance regressions.
          * __Examples__: measuring compile, build, link times for the compiler, individual functions (i.e. data structures sorting, traversing, etc.).
1. **Test category and subcategory**: Tests (sub)categories shall be determined by the project, library, module, and functionality tests are covering.

#### Examples

* F# compiler component test which is verifying generated IL for computation expression will have category `Compiler` and subcategories `EmittedIL` and `ComputationExpressions`.
* F# compiler service unit test which is testing F# tokenizer, will have category `Compiler.Service` and subcategory `Tokenizer`.

Please, refer to [File and project structure](#File-and-project-structure) for more information on how tests will be organized on the filesystem.

### Naming schema

The proposed naming schema for test projects is: `FSharp.Category.Subcategory.TestType`, where
`Category.Subcategory` is either a corresponding source project, or a more generic component (e.g. `Compiler`, `Compiler.Private` or more granular `Compiler.CodeGen`, `Compiler.CodeGen.EmittedIL` if category or subcategory project becomes too big, etc.) and `TestType` is the type of the test (one of `UnitTests`, `ComponentTests`, `IntegrationTests`, `LoadTests`).

### Projects organization

Please refer to the "[Naming schema](#Naming-schema)" section above for more information on the projects naming.

New test projects will be grouped by category and test type, all subcategories are just test folders/files in the test project.

* __Examples__: Having test project organized like:
    > `tests/FSharp.Compiler.ComponentTests/CodeGen/EmittedIL/BasicTests.fs`
    > `tests/FSharp.Compiler.ComponentTests/CodeGen/StringEncoding/StringTests.fs`
    > `tests/FSharp.Compiler.ComponentTests/Optimizations/Inlining/InliningTests.fs`

    Will result in one test dll "`FSharp.Compiler.ComponentTests.dll`" which will contain all the subcategories of tests.
* **Notes**:
  * This will result in reduced fragmentation of tests, all the tests files are under one big category, easier to understand what each component/unit test suite covers, less confusion in test classification for new tests.
  * If some categories (or subcategories) will become big enough - they can be factored out to a separate project.

### Test Utilities/Helpers

For all new and migrated tests, any common/helper functionality shall be factored out to a separate project - `FSharp.Test.Utilities`.

## New tests

* All new tests should be created in the new projects only.
* All new tests should contain a brief docstring description of what is being tested, link to an issue if applicable.
* All new tests should be categorized using xUnit's `Trait`, based on their `Category` and `Subcategories`.

## Migrating existing tests

Existing FSharpQA and Cambridge need to be migrated to corresponding test projects: component-style tests to the `FSharp.Compiler.ComponentTests` and unittest-style tests - `FSharp.Compiler.UnitTests`, `FSharp.Compiler.Private.Scripting.UnitTests`, `FSharp.Build.UnitTests`, etc.

## Next steps

* [**In Progress**] Migrate existing `NUnit` tests to xUnit.
* Clean up CompilerAssert.
* Make PEVerify tests work in netcore/non-windows environment.
* Start migration of existing (namely, FSharpQA and Cambridge) suites to xUnit-based projects.

## Open questions:

* As far as I know, [FSharp.Compiler.Service](https://github.com/fsharp/FSharp.Compiler.Service) is dependant on some of the F# compiler tests. Does it have to be changed as well?

## Other

Related issues: (https://github.com/dotnet/fsharp/issues/7075)

You can find this document under 'tests/README.md'.

### CODE REVIEW

**Comment by jonsequitur** [comment]
It sounds like there's a trade-off between:

* the test framework (runner, mainly?) that just works with the .NET team's infrastructure (e.g. Arcade, `dotnet test`, `.trx` files for reporting results), and,
* a test framework (API, mainly?) that works best with idiomatic F#.

Is there a best-of-both-worlds approach? Are there improvements that we can try to make in XUnit or existing XUnit/F# bridge libraries? Also, does the shift in the F# infrastructure away from .NET Framework and toward .NET Core / .NET 5 change the calculus on anyone's opinions about the testing frameworks we've discussed?

**Comment by KevinRansom** [comment]
@forki  -  you asked about the switch to xunit.  The ci that we use is aware of xunit testing, loging and the like.  Right now we have additional scripts that smooth out some of the differences between the two frameworks within the CI.  We would prefer to get rid of our special cases, and rely on the native CI behaviors.

Kevin

**Comment by KevinRansom** [comment]
@vzarytovskii -  How does language testing fit into to your categorization.  A large portion of our test codebase seems to exist in-order verify that specific language constructs compile and produce executable output.

For example:
Language construct:
     Dotnet Attributes

Type of tests:
    Code generation     Symbols produced

**Comment by KevinRansom** [comment]
@vzarytovskii 

How does language testing fit into the proposed categorization scheme. A large portion of our test codebase seems to exist in-order verify that specific language constructs compile and produce executable output.

For example:
Language construct:
    Dotnet Attributes
    IEnumerable
    For loop

Example types of tests:
    Language conformance
    Code generation
    Symbols production and verification
    Il Verification (peverify), IL producwd
    Use from compiled code
    Use from interactive
    Interop from C#
    Interop with C# implementations

I have no real intuition about what this should look like, but I would appreciate your thoughts on how we might arrange these, so that it is fairly clear where a new test belongs..

**Comment by KevinRansom** [comment]
@vzarytovskii -
forgive me if I overlooked it, but have you considered a continuous testing framework such as NCrunch, should we organise testing inorder to make continuous testing usable and useful if so how would that look?

Thanks

Kevin

**Comment by vzarytovskii** [comment]
@KevinRansom 
> How does language testing fit into the proposed categorization scheme. A large portion of our test codebase seems to exist in-order verify that specific language constructs compile and produce executable output.

> I have no real intuition about what this should look like, but I would appreciate your thoughts on how we might arrange these, so that it is fairly clear where a new test belongs..

My idea is to organize the compiler tests into separate suite(s) of component tests which will be testing certain categories (e.g. codegen, IL r/w, interop, etc.).
The first step is to start moving out of cambridge & FSharpQA suites towards separate xUnit ones, so we can have it runnable/debugable on the standard test platform.

> forgive me if I overlooked it, but have you considered a continuous testing framework such as NCrunch, should we organize testing in order to make continuous testing usable and useful if so how would that look?

To be honest, I did not even consider using NCrunch at this point, due to its licensing (correct me if I'm wrong, but I thought that NC is commercial software, with no OSS license). And I'm unaware of any suitable alternatives which would support both desktop and coreclr.

**Comment by forki** [comment]
From my experience with ncrunch I would not make that a priority. It's an
awesome tool and seeing those red and green bubbles along the covered code
is amazing. But it's really hard to setup properly and to maintain it.

From my experience you get most of the same with just expecto style testing
and dotnet watch run. You don't have the code coverage, but you do have
that constant feedback loop. Which is the important part IMHO

Vlad Zarytovskii <notifications@github.com> schrieb am Do., 11. Juni 2020,
10:55:

**Comment by KevinRansom** [comment]
@vzarytovskii -- As I understand it, there are developers who like to use it as part of their workflow.  Have we considered, whether the changes we make would suit such a workflow?
@colombod  and @jonsequitur are both big fans, so we could start by seeing if they have any advice on how tests should be organized, written and deployed to make continuous testing usefull.

Kevin

**Comment by vzarytovskii** [comment]
@KevinRansom 

> @vzarytovskii -- As I understand it, there are developers who like to use it as part of their workflow.  Have we considered, whether the changes we make would suit such a workflow?

As far as I know, NCrunch is compatible with all modern frameworks and I expect it (and live unittesting) work just fine with xUnit. In fact, after the initial migration it will be a separate goal - to enable these sorts of scenarios for our tests.

**Comment by TIHan** [comment]
We should just simply use `xUnit` and that be our only dependency. At the moment, I don't see value with `FluentAssertions` in the context of testing compilations; most of the assert/verification calls for testing compilations are abstractions that happen to use 'NUnit`/`xUnit` assertions as an implementation detail. But when testing FSharp.Core functionality, it might be a benefit.

Just use `xUnit` and the F# compiler as your only dependencies for compilation/compiler tests. If we find that using other libraries would benefit us, we should should discuss about bringing them in when the time comes. I know `NCrunch` is on the table, but I would not make it a priority.

Also, I strongly encourage everyone to please look at how the Roslyn compiler tests are done. IMO, they make it very simple.

File: tests/FSharp.Test.Utilities/FSharp.Test.Utilities.fsproj [hunk 1]
```diff
@@ -39,6 +40,7 @@
     <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.Common" Version="$(MicrosoftCodeAnalysisWorkspacesCommonVersion)" />
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="$(MicrosoftCodeAnalysisCSharpVersion)" />
     <PackageReference Include="Microsoft.CodeAnalysis.Test.Resources.Proprietary" Version="$(MicrosoftCodeAnalysisTestResourcesProprietaryVersion)" />
-    <PackageReference Include="NUnit" Version="$(NUnitVersion)" />
+    <PackageReference Include="NUnit" Version="$(NUnitVersion)" /> <!-- TODO: This should be removed once all NUnit frameworks are migrated to xUnit -->
+    <PackageReference Include="FluentAssertions" Version="$(FluentAssertionsVersion)" />
```
Comment (TIHan) [inline tests/FSharp.Test.Utilities/FSharp.Test.Utilities.fsproj:44]: Out of curiosity, is `FluentAssertions` giving benefit to any of the current tests?

**Comment by jonsequitur** [comment]
Enabling NCrunch doesn't require that we do anything special in the repo. It's not a library dependency. In practice it works reliably until you overcustomize a build. Keeping the build simple has other virtues anyway, so I think it's worthwhile to keep this kind of workflow within reach.

File: tests/FSharp.Test.Utilities/FSharp.Test.Utilities.fsproj (hunk 1 above)
Comment (vzarytovskii) [reply tests/FSharp.Test.Utilities/FSharp.Test.Utilities.fsproj:44]: I guess only a bit better assertion messages. The seem more readable to me.

**Comment by vzarytovskii** [comment]
I grew with @TIHan here. We should start with plain simple xUnit here and then add other dependencies as needed.

**Comment by zpodlovics** [comment]
Would it be possible to replace the printf "magic" with some [simple formatting functions on the top](https://github.com/dotnet/fsharp/issues/4954#issue-324707426) during the migration ? To avoid issues like this: https://github.com/dotnet/fsharp/issues/4954 as it would help improve the F# test coverage on platforms / runtimes / targets like Mono AOT, WebAssembly, Fable, Fez, etc.

**Comment by vzarytovskii** [comment]
> Would it be possible to replace the printf "magic" with some [simple formatting functions on the top](https://github.com/dotnet/fsharp/issues/4954#issue-324707426) during the migration ? To avoid issues like this: https://github.com/dotnet/fsharp/issues/4954 as it would help improve the F# test coverage on platforms / runtimes / targets like Mono AOT, WebAssembly, Fable, Fez, etc.
> 

Hm, great insight, thanks. I will need to look at it more thoroughly though.

Most likely, won’t  be a part of this particular PR.

File: tests/FSharp.Compiler.UnitTests/ByteMemoryTests.fs [hunk 2]
```diff
@@ -3,14 +3,15 @@ namespace FSharp.Compiler.UnitTests
```
Comment (KevinRansom) [inline tests/FSharp.Compiler.UnitTests/ByteMemoryTests.fs:3]: Should the test files be all in the same directory ````ByteMemoryTests.fs, EditDistance.fs, HashIfExpression.fs.```` or should they be further differentiated, it's just that I imagine eventually Compiler.ComponentTest would have a huge number of files in it.

Comment (vzarytovskii) [reply tests/FSharp.Compiler.UnitTests/ByteMemoryTests.fs:3]: @KevinRansom my idea was grouping files by the directories inside the test suites (e.g. Compiler.ComponentTests). But they can as well be factored out into separate suite if will grow in size.

**Pull request merged**